webapp/src/languages/compiler_flag.rs

```diff
@@ -0,0 +1,152 @@
+    // A flag that consists of a key-value pair
+    KeyValueFlag
+    {
+        key: String,
+        value: String
+    },
+    // A flag that has many values for a single key
+    KeyValuesFlag
+    {
+        key: String,
+        values: HashSet<String>
+    },
+}
+
+impl CompilerFlag
+{
+    pub fn new_single_word(flag: &str) -> CompilerFlag
+    {
+        CompilerFlag::SingleWordFlag(flag.to_owned())
+    }
+
+    pub fn new_key_value(key: &str, value: &str) -> CompilerFlag
+    {
+        CompilerFlag::KeyValueFlag 
+        {
+            key: key.to_owned(),
+            value: value.to_owned()
+        }
+    }
+
+    pub fn new_key_values(key: &str, values: &[&str]) -> CompilerFlag
+    {
+        let mut new_values = HashSet::<String>::new();
+        for value in values
+        {
+            new_values.insert(value.to_string());
+        }
+
+        CompilerFlag::KeyValuesFlag
+        {
+            key: key.to_owned(),
+            values: new_values
+        }
+    }
+}
+
+impl PartialEq for CompilerFlag
+{
+    fn eq(&self, other: &CompilerFlag) -> bool
+    {
+        use CompilerFlag::*;
+        // Compare by keys and values
+        match (self, other)
+        {
+            (Undefined, Undefined) => true,
+            (Undefined, UndefinedFlag(_)) => true,
+            (UndefinedFlag(_), Undefined) => true,
+
+            (UndefinedFlag(self_value), UndefinedFlag(other_value)) => 
+            { 
+                self_value == other_value
+            }
+            (SingleWordFlag(self_flag), SingleWordFlag(other_flag)) =>
+            {
+                self_flag == other_flag
+            },
+            (KeyValueFlag { key: self_key, value: self_value },  
+                KeyValueFlag { key: other_key, value: other_value }) =>
+            {
+                self_key == other_key && self_value == other_value
+            },
+
+            // One to many comparison here is true if a range contains the element it's compared to
+            (KeyValueFlag { key: self_key, value: self_value },  
+                KeyValuesFlag { key: other_key, values: other_values }) =>
+            {
+                self_key == other_key && other_values.contains(self_value)
+            },
+            (KeyValuesFlag { key: self_key, values: self_values },  
+                KeyValueFlag { key: other_key, value: other_value }) =>
+            {
+                self_key == other_key && self_values.contains(other_value)
+            },
+            (KeyValuesFlag { key: self_key, values: self_values },  
+                KeyValuesFlag { key: other_key, values: other_values }) =>
+            {
+                self_key == other_key && self_values == other_values
+            },
+            _ => false
+        }
+    }
+}
+
+impl Hash for CompilerFlag
+{
+    fn hash<H: Hasher>(&self, hasher: &mut H)
+    {
+        // Don't hash by a values of keys
+        match self
+        {   
+            CompilerFlag::Undefined => 
```

Comment:
What are the use cases for `Undefined` and `UndefinedFlag(value)`, why are they two separate types?

Reply:
```Undefined``` is supposed to represent any undefined flag, I plan to use it in ```LangInfo``` flags lists. ```static_info::cpp::construct()``` returns ```LangInfo``` that contains ```Undefined``` in blacklisted flags, which means that any undefined flag is prohibited.
```UndefinedFlag``` also contains values of this flag. To think of it, it's not necessary, but I added it to have a way to return undefined values back to the caller. I honestly find it a little excessive, and it caused issues along the way as I was writing it. Removing it would simplify validation, but I'll have to come up with another way to pass undefined flags' info, if it's even necessary.